v2/apiserver/internal/authz/role_assignments.go

```diff
@@ -0,0 +1,177 @@
+	// ID references a principal. The Type qualifies what type of principal that
+	// is-- for instance, a User or a ServiceAccount.
+	ID string `json:"id,omitempty" bson:"id,omitempty"`
+}
+
+// RoleAssignment represents the assignment of a Role to a principal such as a
+// User or ServiceAccount.
+type RoleAssignment struct {
+	// Role assigns a Role to the specified principal.
+	Role libAuthz.Role `json:"role" bson:"role"`
+	// Principal specifies the principal to whom the Role is assigned.
+	Principal PrincipalReference `json:"principal" bson:"principal"`
+}
+
+// RoleAssignmentsService is the specialized interface for managing
+// RoleAssignments. It's decoupled from underlying technology choices (e.g. data
+// store, message bus, etc.) to keep business logic reusable and consistent
+// while the underlying tech stack remains free to change.
+type RoleAssignmentsService interface {
+	// Grant grants the Role specified by the RoleAssignment to the principal also
+	// specified by the RoleAssignment. If the specified principal does not exist,
+	// implementations must return a *meta.ErrNotFound error.
+	Grant(ctx context.Context, roleAssignment RoleAssignment) error
+
+	// Revoke revokes the Role specified by the RoleAssignment for the principal
+	// also specified by the RoleAssignment. If the specified principal does not
+	// exist, implementations must return a *meta.ErrNotFound error.
+	Revoke(ctx context.Context, roleAssignment RoleAssignment) error
+}
+
+type roleAssignmentsService struct {
+	usersStore           authn.UsersStore
+	serviceAccountsStore authn.ServiceAccountsStore
+	roleAssignmentsStore RoleAssignmentsStore
+}
+
+// NewRoleAssignmentsService returns a specialized interface for managing
+// RoleAssignments.
+func NewRoleAssignmentsService(
+	usersStore authn.UsersStore,
+	serviceAccountsStore authn.ServiceAccountsStore,
+	roleAssignmentsStore RoleAssignmentsStore,
+) RoleAssignmentsService {
+	return &roleAssignmentsService{
+		usersStore:           usersStore,
+		serviceAccountsStore: serviceAccountsStore,
+		roleAssignmentsStore: roleAssignmentsStore,
+	}
+}
+
+func (r *roleAssignmentsService) Grant(
+	ctx context.Context,
+	roleAssignment RoleAssignment,
+) error {
+	switch roleAssignment.Principal.Type {
+	case PrincipalTypeUser:
+		// Make sure the User exists
+		if _, err :=
+			r.usersStore.Get(ctx, roleAssignment.Principal.ID); err != nil {
+			return errors.Wrapf(
+				err,
+				"error retrieving user %q from store",
+				roleAssignment.Principal.ID,
+			)
+		}
+	case PrincipalTypeServiceAccount:
+		// Make sure the ServiceAccount exists
+		if _, err :=
+			r.serviceAccountsStore.Get(ctx, roleAssignment.Principal.ID); err != nil {
+			return errors.Wrapf(
+				err,
+				"error retrieving service account %q from store",
+				roleAssignment.Principal.ID,
+			)
+		}
+	default:
+		return nil
+	}
+
+	// Give them the Role
+	if err := r.roleAssignmentsStore.Grant(ctx, roleAssignment); err != nil {
```

Comment:
If we wanted, we could simplify lines 111-122 to:
```
return errors.Wrapf(r.roleAssignmentsStore.Grant(ctx, roleAssignment), "error granting role..."...
```

(Similarly for Revoke below.)

No opinion either way; disregard if you'd already thought of it but prefer this version.

Reply:
I always forget that there's no harm wrapping a nil error. Can do.

Reply:
Went to apply this and it looks like this:

```
// Give them the Role
return errors.Wrapf(
  r.roleAssignmentsStore.Grant(ctx, roleAssignment),
  "error granting role %q with scope %q to %s %q in store",
  roleAssignment.Role.Name,
  roleAssignment.Role.Scope,
  roleAssignment.Principal.Type,
  roleAssignment.Principal.ID,
)
```

Invoking "clear is better than clever," I do think I prefer how it was before.

Reply:
Ah yea, that's quite a return statement!  Okay, I think I do, too!  Thanks for taking a look.

Reply:
I'm less concerned with it being big and more concerned that a very important function call is buried inside what might be dismissed as "just error-handling code."